neonvm/runner/main.go

```diff
@@ -779,6 +786,47 @@ func listenForCPUChanges(ctx context.Context, logger *zap.Logger, port int32, cg
 	}
 }
 
+// writes from socket line by line
+func forwardLogs(ctx context.Context, logger *zap.Logger, wg *sync.WaitGroup) error {
```

Comment:
Some of the control flow in this function seems undesirable.

As far as I can tell:

- If there's no logs from the VM, we wait for up to 3s after termination to decide we're done (corollary: This will increase runner pod termination time [and therefore suspend_compute time], by up to 3 seconds — may be fine in practice, but worth being confident that it's fine)
- If we get an error reading from the socket, we retry 3s later (IMO we should have exponential backoff here, with a _limit_ of a few seconds)
- If the context expires while we're reading log lines, then we will drop the rest (probably not what we want to do?), leaving the logs from QEMU shutdown unread.
    - imo: This is ok if we get an error. But if we're still reading logs, we should print those!

Not 100% sure on these, please correct me if I made a mistake.

It'd also be good to know what QEMU's behavior around the socket is when the VM exits — e.g. do we get a particular kind of error?